src/tagsQLController.js

```diff
@@ -0,0 +1,395 @@
+// "Natural language" operators
+let OPERATOR_EQ = '=';
+let OPERATOR_NOTEQ = '!=';
+let OPERATOR_IN = 'is in';
+let OPERATOR_NOTIN = 'is not in';
+let OPERATOR_EXISTS = 'exists';
+let OPERATOR_NOTEXISTS = 'doesn\'t exist';
+let OPERATOR_AND = 'AND';
+let OPERATOR_OR = 'OR';
+
+export class TagsQLController {
+
+  constructor(uiSegmentSrv, datasource, $q, targetSupplier)  {
+    this.uiSegmentSrv = uiSegmentSrv;
+    this.datasource = datasource;
+    this.$q = $q;
+    this.targetSupplier = targetSupplier;
+    this.removeTagSegment = uiSegmentSrv.newSegment({fake: true, value: '-- Remove tag --'});
+    this.removeValueSegment = uiSegmentSrv.newSegment({fake: true, value: '-- Remove value --'});
+  }
+
+  initTagsSegments() {
+    let target = this.targetSupplier();
+    if (!target.tagsQL && target.tags) {
+      // Compatibility, switching from older version
+      target.tagsQL = convertFromKVPairs(target.tags);
+    }
+    var segments = stringToSegments(target.tagsQL, this.uiSegmentSrv);
+    segments.push(this.uiSegmentSrv.newPlusButton());
+    // Fix plus-button: add it at the end of each enumeration
+    var isInEnum = false;
+    for (var i = 0; i < segments.length; i++) {
+      if (segments[i].type === 'operator' && (segments[i].value === OPERATOR_IN || segments[i].value === OPERATOR_NOTIN)) {
+        isInEnum = true;
+      } else if (isInEnum && segments[i].type !== 'value') {
+        segments.splice(i, 0, this.uiSegmentSrv.newPlusButton());
+        isInEnum = false;
+      }
+    }
+    return segments;
+  }
+
+  getTagsSegments(segments, segment, $index) {
+    // Get suggestions for available values in a given segment
+    if (segment.type === 'condition') {
+      return this.$q.when([this.uiSegmentSrv.newSegment(OPERATOR_AND), this.uiSegmentSrv.newSegment(OPERATOR_OR)]);
+    }
+    if (segment.type === 'operator') {
+      return this.$q.when(this.uiSegmentSrv.newOperators([OPERATOR_EQ, OPERATOR_NOTEQ, OPERATOR_EXISTS, OPERATOR_NOTEXISTS, OPERATOR_IN, OPERATOR_NOTIN]));
+    }
+    if (segment.type === 'plus-button') {
+      // Find previous operator to know if we're in an enumeration
+      var i = this.getContainingEnum(segments, $index);
+      if (i > 0) {
+        let key = segments[i-1].value;
+        return this.datasource.suggestTags(this.targetSupplier().type, key)
+          .then(this.uiSegmentSrv.transformToSegments(false));
+      } else {
+        return this.getTagKeys();
+      }
+    } else if (segment.type === 'key')  {
+      return this.getTagKeys()
+          .then(keys => [angular.copy(this.removeTagSegment)].concat(keys));
+    } else if (segment.type === 'value')  {
+      // Find preceding key
+      var i = $index - 2;
+      while (segments[i].type !== 'key') {
+        i--;
+      }
```

Comment:
Here it looks for the key that precedes the current index. 'findIndex' apparently can search from a given index, but incrementally ; I need reverse order here.